Automattic/wp-calypso
# Remove localforage from Calypso.

This change implements the new `lib/browser-storage`, which uses `IndexedDB` when available, falling back to `localStorage` otherwise.

It also removes `localforage` and `lib/localforage`, a utility library around `localforage`.

This change should have a positive impact on our bundle size, by removing `localforage` (around 8KB compressed) and replacing it with this much smaller codebase.

@jsnajdr, @blowery: Please take a look at the implementation and let me know what you think. ~Note that I haven't added support for feature-detecting broken `IndexedDB` implementations, but I'm happy to do that if you feel it's important.~ Edit: now added.

@brandonpayton: Please ensure that this change doesn't break support sessions, in that there is still no cross-contamination between users.

#### Changes proposed in this Pull Request

* Remove `localforage` dependency.
* Remove `lib/localforage`.
* Implement new `lib/browser-storage`.
* Update all tests and code accordingly.

#### Testing instructions

General usage:
Ensure Redux state is stored and retrieved correctly. Namely:
* A previously saved Redux state (using the `localforage` implementation) is correctly read using this implementation.
* This implementation doesn't produce any errors if the store is empty on load (i.e., there is no previously-stored data).

Support sessions:
* Ensure that bypassing persistent state works correctly, with no cross-contamination.

Fixes #28719
